File: MatlabCodes/Chapter 11/Fig11_06.m
function CIR_IR_paths
clc;close all;

% CIR model parameter settings

NoOfPaths = 1; 
NoOfSteps = 500; 
T         = 50.0;
lambda     = 0.1;
gamma     = 0.1;
r0        = 0.05;
theta     = 0.05;

% Effect of lambda on Monte Carlo paths

figure(1); hold on; grid on;
lambdaV = [0.01, 0.2, 5.0];
argLegend = cell(3,1);
idx = 1;
for lambdaTemp= lambdaV
    randn('seed',7)
    [R,timeGrid] = GeneratePathsCIREuler(NoOfPaths,NoOfSteps,r0,lambdaTemp,theta,gamma,T);
    plot(timeGrid,R,'linewidth',1.5)
    argLegend{idx} = sprintf('lambda=%.2f',lambdaTemp);
    idx = idx + 1;
end
legend(argLegend)
xlim([0,T])
xlabel('time')
ylabel('r(t)')

% Effect of eta on Monte Carlo paths

figure(2); hold on; grid on;
gammaV = [0.1, 0.2, 0.3];
argLegend = cell(3,1);
idx = 1;
for gammaTemp= gammaV
    randn('seed',1)
    [R,timeGrid] = GeneratePathsCIREuler(NoOfPaths,NoOfSteps,r0,lambda,theta,gammaTemp,T);
    plot(timeGrid,R,'linewidth',1.5)
    argLegend{idx} = sprintf('gamma=%.2f',gammaTemp);
    idx = idx + 1;
end
legend(argLegend)
xlim([0,T])
xlabel('time')
ylabel('r(t)')

function [R,time] = GeneratePathsCIREuler(NoOfPaths,NoOfSteps,r0,lambda,theta,gamma,T)

% Define initial value

R=zeros(NoOfPaths,NoOfSteps);
R(:,1) = r0;

% Random noise

Z=random('normal',0,1,[NoOfPaths,NoOfSteps]);
W=zeros([NoOfPaths,NoOfSteps]);

dt = T / NoOfSteps;
time = zeros([NoOfSteps+1,1]);
for i=1:NoOfSteps
    if NoOfPaths>1
        Z(:,i)   = (Z(:,i) - mean(Z(:,i))) / std(Z(:,i));
    end
    W(:,i+1)  = W(:,i) + sqrt(dt).*Z(:,i);
    R(:,i+1)  = R(:,i) + lambda*(theta-R(:,i))*dt+  gamma* sqrt(R(:,i)).* (W(:,i+1)-W(:,i));
        
    % We apply here the truncation scheme to deal with negative values

    R(:,i+1) = max(R(:,i+1),0);
    time(i+1) = time(i) + dt;
end

File: PythonCodes/Chapter 11/Fig11_06.py
#%%
"""
Created on Jan 23 2019
Paths for the CIR model
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st
import scipy.integrate as integrate
from mpl_toolkits import mplot3d

def GeneratePathsCIREuler(NoOfPaths,NoOfSteps,T,lambd,r0,theta,gamma):    
    Z = np.random.normal(0.0,1.0,[NoOfPaths,NoOfSteps])
    W = np.zeros([NoOfPaths, NoOfSteps+1])
    R = np.zeros([NoOfPaths, NoOfSteps+1])
    R[:,0]=r0
    time = np.zeros([NoOfSteps+1])
        
    dt = T / float(NoOfSteps)
    for i in range(0,NoOfSteps):

        # Making sure that samples from a normal have mean 0 and variance 1

        if NoOfPaths > 1:
            Z[:,i] = (Z[:,i] - np.mean(Z[:,i])) / np.std(Z[:,i])
        W[:,i+1] = W[:,i] + np.power(dt, 0.5)*Z[:,i]
        R[:,i+1] = R[:,i] + lambd*(theta - R[:,i]) * dt + gamma* np.sqrt(R[:,i]) * (W[:,i+1]-W[:,i])

        # Truncated boundary condition

        R[:,i+1] = np.maximum(R[:,i+1],0.0)
        time[i+1] = time[i] +dt
        
    # Outputs

    paths = {"time":time,"R":R}
    return paths
def mainCalculation():
    NoOfPaths = 1
    NoOfSteps = 500
    T         = 50.0
    lambd     = 0.1
    gamma     = 0.05
    r0        = 0.05
    theta     = 0.05
    
    # Effect of mean reversion lambda

    plt.figure(1) 
    legend = []
    lambdVec = [0.01, 0.2, 5.0]
    for lambdTemp in lambdVec:    
        np.random.seed(1)
        Paths = GeneratePathsCIREuler(NoOfPaths,NoOfSteps,T,lambdTemp,r0,theta,gamma)
        legend.append('lambda={0}'.format(lambdTemp))
        timeGrid = Paths["time"]
        R = Paths["R"]
        plt.plot(timeGrid, np.transpose(R))   
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("R(t)")
    plt.legend(legend)
        
    # Effect of the volatility

    plt.figure(2)    
    legend = []
    gammaVec = [0.1, 0.2, 0.3]
    for gammaTemp in gammaVec:
        np.random.seed(1)
        Paths = GeneratePathsCIREuler(NoOfPaths,NoOfSteps,T,lambd,r0,theta,gammaTemp)
        legend.append('gamma={0}'.format(gammaTemp))
        timeGrid = Paths["time"]
        R = Paths["R"]
        plt.plot(timeGrid, np.transpose(R))   
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("R(t)")
    plt.legend(legend)
    
mainCalculation()
